nixos/modules/services/networking/amazon-ec2-net-utils.nix

```diff
@@ -0,0 +1,80 @@
+{
+  lib,
+  pkgs,
+  config,
+  ...
+}:
+let
+  cfg = config.services.amazon-ec2-net-utils;
+in
+{
+  options.services.amazon-ec2-net-utils = {
+    enable = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        Whether to enable amazon-ec2-net-utils.
+
+        This service relies on systemd-networkd so `systemd.network.enable` must also be `true`.
+      '';
+      example = true;
+    };
+    package = lib.mkPackageOption pkgs "amazon-ec2-net-utils" { };
+  };
+
+  config =
+    lib.mkIf cfg.enable
+    # amazon-ec2-net-utils needs systemd-networkd.
+    && config.systemd.network.enable {
+      # See https://github.com/amazonlinux/amazon-ec2-net-utils/tree/v2.5.1/systemd/system.
+      systemd = {
+        services = {
+          "refresh-policy-routes@" = {
+            description = "Refresh policy routes for %I";
+            serviceConfig = {
+              Type = "oneshot";
+              PrivateTmp = "yes";
+              AmbientCapabilities = "CAP_NET_ADMIN";
+              NoNewPrivileges = "yes";
+              User = "root";
+              ExecStart = "${cfg.package}/bin/setup-policy-routes %i refresh";
+              SuccessExitStatus = "SIGTERM";
+              KillMode = "process";
+            };
+          };
+
+          "policy-routes@" = {
+            description = "Set up policy routes for %I";
+            startLimitIntervalSec = 10;
+            startLimitBurst = 5;
+            wants = [ "refresh-policy-routes@%i.timer" ];
+            # TODO: Need [Install] for Also = "refresh-policy-routes@%i.timer". systemd.services.<name> has no installConfig attribute.
```

Comment:
@commiterate `[Install]` does absolutely nothing on NixOS. That's why you'll see `wantedBy` defined in most services in nixos modules. We populate `[Install]` with the contents of the `wantedBy` option and its friends, but that's largely for illustrative purposes; we don't actually use the `[Install]` section. We've thought about changing that, but it hasn't been done yet.

Reply:
Even the `[Install]` sections of units from `systemd.packages` are ignored in NixOS. But, AFAICT, the `[Install]` sections of those upstream units are doing absolutely nothing anyway. The only thing that `[Install]` does is create `Wants=` dependencies and its friends by creating `/etc/systemd/system/*.wants/*` symlinks when `systemctl enable` is run. e.g. Having `WantedBy=foo.service` in `bar.service` means that running `systemctl enable bar.service` will create `/etc/systemd/system/foo.service.wants/bar.service -> ../bar.service`.

Of course this isn't a thing on NixOS; we create those statically during system build via the `systemd.$unittype.$name.wantedBy` options and its friends. The `Also=` directive just means that when you run `systemctl enable bar.service`, it also does the `enable` operation for anything in `Also=`. Given that *none* of the units in the upstream `amazon-ec2-net-utils` source tree do anything like `WantedBy=` in their `[Install]` sections, that means it all does effectively nothing.

Now, I've just realized that what that udev rule does is *marvelously stupid*. It runs `systemctl enable/disable --now` on those units. This is really poorly done. First of all, as I've just discussed, the `enable` operation will have no effect for these units normally. But if it did, it wouldn't work on NixOS thanks to the immutable store and `/etc/systemd/system/`. But the `--now` flag causes the units to be started/stopped too. So *really* that udev rule should be using `start` and `stop`, not `enable` and `disable`.

Wow. Really badly done by the amazon folks. `enable`/`disable` is supposed to be an administrator's decision, not the vendor's.

You'll want to fix the rule to not be stupid, and then merely adding the upstream units to `systemd.package` (with patched paths) will work.

Reply:
Yes, presets are the correct thing to be using on other distros. Though, as I explained, it still wouldn't even do anything since there's no actual content in the `[Install]` sections :P